### ISSUE TITLE
Add dotnet template to create a package that follows best practices

### ISSUE THREAD

**Issue by loic-sharma** [NuGet/Home#10956]
When creating a NuGet package there are several steps you need to do to follow best practices: https://docs.microsoft.com/en-us/nuget/create-packages/package-authoring-best-practices

While this doc contains all the information you need, some folks may prefer to use a .NET template to bootstrap their project:

```
dotnet new nuget
```

We could add options too:

```
dotnet new nuget --sourcelink GitHub
```

The generated project would guide users to follow best practices through ample documentation. For example:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net5.0</TargetFramework>
  </PropertyGroup>

  <!-- NuGet package metadata -->
  <!-- See: https://docs.microsoft.com/nuget/create-packages/package-authoring-best-practices -->
  <PropertyGroup>
    <PackageIcon>packageIcon.png</PackageIcon>
    <PackageId>my-package</PackageId>
    <PackageLicenseExpression>MIT</PackageLicenseExpression>
    <PackageTags>tag1; tag2</PackageTags>

    <!-- And so on... -->
  </PropertyGroup>

  <ItemGroup>
    <None Include="packageIcon.png" Pack="true" PackagePath=""/>
  </ItemGroup>

</Project>
```

(This project file is entirely hypothetical. We should spend some quality time to decide what should be included)

Credit goes to @bruno-garcia for the idea

**Comment by bruno-garcia** [comment]
An example README (referred to by the csproj) would be nice too

**Comment by zivkan** [comment]
The NuGet team doesn't "own" the dotnet CLI's templates, even the nugetconfig template [appears to have been added without discussion/notification to us](https://github.com/dotnet/templating/commit/9f65552dea065bf486da38ec569d19e5cdc9ce33#diff-6bd932836569360572b641db949c136fa5be5afa712cccd2e5d71a7838d14130).  So, I propose that this issue belongs in dotnet/templating, where those templates live. Unfortunately github doesn't allow issues to be transferred across org boundaries, and I don't see any other built-in way to transfer it, so I propose that you re-create the issue there.


Secondly, rather than a new template named `nuget`, I think it would be better to have an argument `dotnet new classlib --nuget-defaults`, or `--as-package` or something like that (I would expect the vast majority of projects being packed are classlibs, hence would not have `<OutputType>exe` as in the original comment).  For classlibs it could add the defaults for packages, but for console apps it could add defaults to correctly pack as a dotnet tool.

**Comment by zkat** [comment]
We need to be more data-driven about this, I think: We should focus on specific things we want to improve, have metrics to measure, and then decide if this is something that can work towards that.

Per Nikolche, we think this is the sort of thing @JonDouglas' nuget scorer tool can help us start measuring, so we can figure out exactly what we want to do going forward. How does that sound? In the meantime, we're gonna icebox this idea (instead of closing it). We should definitely revisit it once Jon's tool is ready, though.

**Comment by loic-sharma** [comment]
Surveys and customer development show that customers find it difficult to create high quality packages. [For example](https://twitter.com/buhakmeh/status/1407699970476609540?s=20):

![image](https://user-images.githubusercontent.com/737941/123315655-cce11d00-d4e0-11eb-8bd2-379f859079b4.png)

We documented [authoring best practices](https://docs.microsoft.com/en-us/nuget/create-packages/package-authoring-best-practices) as it was a low hanging fruit to help our authors. Adding a template is yet another low hanging fruit that would also help.

> We need to be more data-driven about this ... We should focus on specific things we want to improve, have metrics to measure

I complete agree! We do have data on our ecosystem here: https://github.com/nuget/insights

Here are areas we could *definitely* improve that we have metrics we can measure:

* Fix privacy and security concerns by reducing usage of the deprecated `IconUrl`
* Improve the percentage of packages that are debuggable
* Improve the percentage of packages whose build are reproducible
* Drive adoption of the new README feature. It is currently *very* low

> Per Nikolche, we think this is the sort of thing @JonDouglas' nuget scorer tool can help us start measuring, so we can figure out exactly what we want to do going forward. How does that sound?

While I agree package scoring would help, it is a large effort that will take several sprints to implement. Given the server team's backlog, we may not be able to start this work until 2022 at the earliest. On the other hand, I expect that defining and creating a new .NET template would be significantly cheaper (possibly a week of work as a guesstimate).

**Comment by jeromelaban** [comment]
@loic-sharma when talking about reproducible, you're talking about adding a reference to https://github.com/dotnet/reproducible-builds ?

**Comment by loic-sharma** [comment]
@jeromelaban Yup. Ideally all new packages uploaded to nuget.org would be green on [nuget.info's `Health` section](https://nuget.info/packages/Newtonsoft.Json/13.0.1) :)

![image](https://user-images.githubusercontent.com/737941/123318467-141cdd00-d4e4-11eb-8af8-8fb5fe3b100a.png)

**Comment by zkat** [comment]
@JonDouglas @anangaur could we maybe make an epic that covers the original proposal for this issue, as well as more issues for the various metric improvements that @loic-sharma suggested?
